moshi/src/test/java/com/squareup/moshi/JsonAdapterTest.java

```diff
@@ -182,4 +183,22 @@
     serializeNulls.toJson(writer, Collections.<String, String>singletonMap("a", null));
     assertThat(factory.json()).isEqualTo("{\"a\":null}");
   }
+
+  @Test public void stringDocumentMustBeFullyConsumed() throws IOException {
+    JsonAdapter<String> brokenAdapter = new JsonAdapter<String>() {
+      @Override public String fromJson(JsonReader reader) throws IOException {
+        return "Forgot to call reader.nextString().";
+      }
+
+      @Override public void toJson(JsonWriter writer, @Nullable String value) throws IOException {
+        throw new AssertionError();
+      }
+    };
+    try {
+      brokenAdapter.fromJson("\"value\"");
```

Comment:
This is actually somewhat of a surprising test. I’d expect a regular JsonAdapter and input of like `true true` or something

Reply:
do you have a test case you'd like to see?
making the new check only for non-lenient adapters will mean `true true` will either (in a strict adapter) fail as invalid JSON or (in a lenient adapter) won't be checked.

Reply:
Yeah I think that's exactly how the test should look.

Reply:
roger. I left this test because it's the only case where the message `"JSON document was not fully consumed."` can ever be seen. I updated with `true true` tests.